src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -289,68 +335,122 @@ private static void assertVersionMatch() {
+
+        return tempResourceFilePath;
+    }
+
     /**
-     * Unfortunately, we cannot actually use Files.createTempFile, because that internally depends on
-     * SecureRandom, which results in a circular dependency. So, for now, we just read from
-     * /dev/urandom. Clearly this won't work when we start supporting windows systems. We are intentionally taking
-     * as few dependencies here as possible.
+     * We need a source of entropy to create a random temporary directory name at startup before we've loaded native
+     * crypto libraries. So, for now, we just read from /dev/urandom. Clearly this won't work when we start supporting
+     * Windows systems. We are intentionally taking as few dependencies here as possible.
      */
-    private static synchronized Path createTmpFile(final String prefix, final String suffix) throws IOException {
+    private static byte[] bootstrapRng(int numBytes) throws IOException {
         final Path urandomPath = Paths.get("/dev/urandom");
         if (!Files.exists(urandomPath)) {
             throw new AssertionError("/dev/urandom must exist for bootstrapping");
         }
-        final Path tmpDir = Paths.get(System.getProperty("java.io.tmpdir"));
-        if (!Files.isDirectory(tmpDir)) {
-            throw new AssertionError("java.io.tmpdir is not valid: " + tmpDir);
-        }
 
-        final FileAttribute<Set<PosixFilePermission>> permissions =
-                PosixFilePermissions.asFileAttribute(new HashSet<>(Arrays.asList(
-                    PosixFilePermission.OWNER_READ,
-                    PosixFilePermission.OWNER_WRITE,
-                    PosixFilePermission.OWNER_EXECUTE
-        )));
-
-        final byte[] rndBytes = new byte[Long.BYTES]; // Default java tmp files use this much entropy
-        final int RETRY_LIMIT = 1000;
+        final byte[] rndBytes = new byte[numBytes];
+        final int RETRY_LIMIT = 10;
         try (InputStream rndStream = Files.newInputStream(urandomPath, StandardOpenOption.READ)) {
             int attempt = 0;
             // We keep doing this until we can create something new or fail badly
             while (attempt < RETRY_LIMIT) {
                 attempt++;
-                if (rndStream.read(rndBytes) != rndBytes.length) {
-                    throw new AssertionError("Unable to read enough entropy");
+                if (rndStream.read(rndBytes) == rndBytes.length) {
+                    return rndBytes;
                 }
+            }
+        }
+        throw new AssertionError("Unable to read enough entropy");
+    }
+
+    /**
+     * Unfortunately, we cannot actually use Files.createTempFile, because that internally depends on
+     * SecureRandom, which results in a circular dependency.
+     */
+    private static synchronized Path createPrivateTmpDir(final String prefix) throws IOException {
+        final Path systemTempDir = Paths.get(System.getProperty("java.io.tmpdir"));
```

Comment:
is this property always guaranteed to be present? if not, does `Files.isDirectory` handle `null` well or does it blow up w/ NPE?

Reply:
`Paths.get()` will return `java.nio.file.InvalidPathException` if `java.io.tmpdir` is not set. However this was the previously existing behavior before this PR, so I think it is acceptable.

Reply:
ah, yep looking at tip-of-`develop`, i see that we'd already assumed the presence of that property. fair enough.